src/rviz/default_plugin/marker_utils.cpp

```diff
@@ -76,4 +84,289 @@ MarkerBase* createMarker(int marker_type, MarkerDisplay* owner, DisplayContext*
+{
+  std::stringstream ss;
+
+  if(marker.action != visualization_msgs::Marker::ADD)
+    return true;
+
+  switch (marker.type) {
+  case visualization_msgs::Marker::ARROW:
+    ss << checkQuaternion(marker);
+    ss << checkScale(marker);
+    ss << checkColor(marker);
+    ss << checkPointsArrow(marker);
+    ss << checkColorsEmpty(marker);
+    ss << checkTextEmpty(marker);
+    ss << checkMeshEmpty(marker);
+    break;
+
+  case visualization_msgs::Marker::CUBE:
+  case visualization_msgs::Marker::CYLINDER:
+  case visualization_msgs::Marker::SPHERE:
+    ss << checkQuaternion(marker);
+    ss << checkScale(marker);
+    ss << checkColor(marker);
+    ss << checkPointsEmpty(marker);
+    ss << checkColorsEmpty(marker);
+    ss << checkTextEmpty(marker);
+    ss << checkMeshEmpty(marker);
+    break;
+
+
+  case visualization_msgs::Marker::LINE_STRIP:
+  case visualization_msgs::Marker::LINE_LIST:
+    ss << checkQuaternion(marker);
+    ss << checkScaleLineStripAndList(marker);
+    ss << checkPointsNotEmpty(marker);
+    ss << checkColors(marker);
+    ss << checkTextEmpty(marker);
+    ss << checkMeshEmpty(marker);
+    break;
+
+  case visualization_msgs::Marker::SPHERE_LIST:
+  case visualization_msgs::Marker::CUBE_LIST:
+  case visualization_msgs::Marker::TRIANGLE_LIST:
+    ss << checkQuaternion(marker);
+    ss << checkScale(marker);
+    ss << checkPointsNotEmpty(marker);
+    ss << checkColors(marker);
+    ss << checkTextEmpty(marker);
+    ss << checkMeshEmpty(marker);
+    break;
+
+  case visualization_msgs::Marker::POINTS:
+    ss << checkScalePoints(marker);
+    ss << checkPointsNotEmpty(marker);
+    ss << checkColors(marker);
+    ss << checkTextEmpty(marker);
+    ss << checkMeshEmpty(marker);
+    break;
+
+  case visualization_msgs::Marker::TEXT_VIEW_FACING:
+    ss << checkColor(marker);
+    ss << checkScaleText(marker);
+    ss << checkTextNotEmptyOrWhitespace(marker);
+    ss << checkPointsEmpty(marker);
+    ss << checkColorsEmpty(marker);
+    ss << checkMeshEmpty(marker);
+    break;
+
+  case visualization_msgs::Marker::MESH_RESOURCE:
+    ss << checkQuaternion(marker);
+    ss << checkColor(marker);
+    ss << checkScale(marker);
+    ss << checkPointsEmpty(marker);
+    ss << checkColorsEmpty(marker);
+    ss << checkTextEmpty(marker);
+
+    break;
+
+  default:
+    ss << "Unknown marker type: " << marker.type ;
+  }
+
+  std::string warning = ss.str();
+  if(!warning.empty())
+  {
+    if(warning.substr(warning.length()-2) == ", ")
+      warning = warning.erase(warning.length()-2, 2);
+
+    owner->setMarkerStatus(MarkerID(marker.ns, marker.id), StatusProperty::Warn, warning);
+    ROS_WARN("Marker '%s/%d': %s", marker.ns.c_str(), marker.id, warning.c_str());
+    return false;
+  }
+
+  owner->deleteMarkerStatus(MarkerID(marker.ns, marker.id));
+  return true;
+}
+
+bool checkMarkerArrayMsg(const visualization_msgs::MarkerArray& array, MarkerDisplay* owner)
+{
+  std::set<MarkerID> marker_set;
```

Comment:
Please use a **sorted** `std::vector` here, reserving `array.markers.size()` elements. [`std::set` is rather costly](http://lafstern.org/matt/col1.pdf) and not needed here. You can easily check for duplicate items using `std::lower_bound` as indicated in the reference.

Reply:
Set has been replaced as suggested by a `std::vector` in 08c8351